File: Design/146.LRU-Cache/146.LRU-Cache.ts
class LinkedListNode {
  public val: number;
  public prev: LinkedListNode | null;
  public next: LinkedListNode | null;
  public key: number;
  constructor(
    key: number,
    val: number,
    prev?: LinkedListNode,
    next?: LinkedListNode
  ) {
    this.val = val;
    this.key = key;
    this.prev = prev === undefined ? null : prev;
    this.next = next === undefined ? null : next;
  }
}
class LRUCache {
  private readonly sentinel: LinkedListNode;
  private capacity: number;
  private nodeMap: Map<number, LinkedListNode>;
  public constructor(capacity: number) {
    this.sentinel = new LinkedListNode(0, 0);
    this.sentinel.prev = this.sentinel;
    this.sentinel.next = this.sentinel;
    this.capacity = capacity;
    this.nodeMap = new Map();
  }

  get(key: number): number {
    const node = this.nodeMap.get(key);

    if (node === undefined) {
      return -1;
    }

    // hoisting to last of the linkedlist
    if (this.sentinel.prev !== node) {
      this.remove(node);
      this.addLast(node.key, node.val);
    }

    return node.val;
  }

  put(key: number, value: number): void {
    const node = this.nodeMap.get(key);

    if (node) {
      this.remove(node);
      this.addLast(key, value);
      return;
    }

    if (this.nodeMap.size === this.capacity) {
      this.remove(this.sentinel.next);
    }
    this.addLast(key, value);
  }

  private remove(node: LinkedListNode): void {
    if (node === this.sentinel.next) {
      this.sentinel.next = this.sentinel.next.next;
      this.sentinel.next.prev = this.sentinel;
    } else {
      node.prev.next = node.next;
      node.next.prev = node.prev;
    }
    this.nodeMap.delete(node.key);
  }

  private addLast(key: number, val: number): void {
    const node = new LinkedListNode(
      key,
      val,
      this.sentinel.prev,
      this.sentinel
    );
    this.sentinel.prev.next = node;
    this.sentinel.prev = node;
    this.nodeMap.set(key, node);
  }
}

File: Design/146.LRU-Cache/146.LRU-Cache.py
class Node:

    def __init__(self, key, value, prev = None, next = None):
        self.key = key
        self.value = value
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.currentSize = 0
        self.sentinel = Node(0,0)
        self.sentinel.prev = self.sentinel
        self.sentinel.next = self.sentinel
        self.notebook = dict()
        

    def get(self, key: int) -> int:
      notebook = self.notebook
      if key in notebook:
        node = notebook[key]
        self.deleteNode(node)
        self.addNodeToFirst(node)
        return node.value
      return -1

    def put(self, key: int, value: int) -> None:
      if key in self.notebook:
         node = self.notebook[key]
         node.value = value
         self.deleteNode(node)
         self.addNodeToFirst(node)
         return

      if self.currentSize == self.capacity:
        lastNode = self.sentinel.prev
        self.deleteNode(lastNode)
        del self.notebook[lastNode.key]
        self.currentSize -= 1

      node = Node(key, value)
      self.addNodeToFirst(node)
      self.notebook[key] = node
      self.currentSize += 1
      
    def addNodeToFirst(self, node: Node):
      next = self.sentinel.next
      node.prev = self.sentinel
      node.next = next
      next.prev = node
      self.sentinel.next = node

    def deleteNode(self, node: Node):
      prev = node.prev
      next = node.next
      prev.next = next
      next.prev = prev
      node.next = None
      node.prev = None


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
